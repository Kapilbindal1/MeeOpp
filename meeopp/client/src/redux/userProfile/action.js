import action from '../action';
// Constant
export const USER_PROFILE_DETAILS = 'user_profile_details';
export const GET_USER_PROFILE_DETAILS = 'get_user_profile_details';
// Action to call to a backend set user details API
export const userProfileDetails = data =>
  action({
    type: 'USER_PROFILE_DETAILS',
    payload: {
      request: {
        url: '/userProfileDetails',
        method: 'POST',
        data,
      },
    },
  });

// Action to call to a backend get user details  API
export const getUserDetails = data =>
  action({
    type: 'GET_USER_PROFILE_DETAILS',
    payload: {
      request: {
        url: '/getUserDetails',
        method: 'POST',
        data,
      },
    },
  });
