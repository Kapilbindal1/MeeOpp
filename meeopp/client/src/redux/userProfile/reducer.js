const defaultState = {
  firstName: '',
  lastName: '',
  company: '',
  department: '',
  position: '',
  email: '',
  id: '',
};

// Function to return updated state based on responce of API's
const setUserProfileDetails = (state = defaultState, action) => {
  if (action && action.payload && action.payload && action.payload.data) {
    const data = action.payload.data.user;
    return {
      ...state,
      firstName: data.firstname,
      lastName: data.lastname,
      company: data.company,
      department: data.department,
      position: data.position,
      email: data.email,
      // eslint-disable-next-line no-underscore-dangle
      id: data._id,
    };
  }
  return {};
};

const getFormattedMessage = action => {
  const message = 'Internal server error.';
  if (action && action.error) {
    if (action.error.response) {
      const { response } = action.error;
      if (response && response.data && response.data.message) {
        return response.data.message;
      }
    } else if (action.error.data) {
      const { data } = action.error;
      if (typeof data === 'string') {
        return data;
      }
    }
  }
  return message;
};

const handleFailedCase = action => {
  const actionType = action.type;
  if (actionType && actionType.toUpperCase().includes('_FAIL')) {
    const message = getFormattedMessage(action);
    // eslint-disable-next-line no-console
    console.log('message: ', message);
    return message;
  }
  return null;
};

// Reducer for setting and getting user profile details (database)
const setUserProfileDetailsReducer = (state = defaultState, action) => {
  switch (action.type) {
    case 'USER_PROFILE_DETAILS_SUCCESS': {
      const updatedState = setUserProfileDetails(state, action);
      return {
        ...updatedState,
      };
    }
    case 'GET_USER_PROFILE_DETAILS_SUCCESS': {
      const updatedState = setUserProfileDetails(state, action);
      return {
        ...updatedState,
      };
    }
    case 'GET_USER_PROFILE_DETAILS_FAIL': {
      const failedResult = handleFailedCase(action);
      return {
        failedResult,
      };
    }
    case 'USER_PROFILE_DETAILS_FAIL': {
      const failedResult = handleFailedCase(action);
      return {
        failedResult,
      };
    }
    default:
      return state;
  }
};

export default setUserProfileDetailsReducer;
