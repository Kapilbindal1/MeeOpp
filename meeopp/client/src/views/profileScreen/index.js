/* eslint-disable no-alert */
/* eslint-disable no-undef */
import React, { Component } from 'react';
import { Card, CardBody, Button, Label } from 'reactstrap';
import { connect } from 'react-redux';
import '../../App.css';
import PropTypes from 'prop-types';
import logo from '../../assets/meeopplogo.png';
import { userProfileDetails, getUserDetails } from '../../redux/userProfile/action';
import Validation from '../validations';

export class ProfileScreen extends Component {
  constructor(props) {
    super(props);
    const { firstName, lastName, company, department, position, email, id } = props.allDetails;
    this.state = {
      firstName,
      lastName,
      company,
      department,
      position,
      email,
      // eslint-disable-next-line react/no-unused-state
      id,
    };
    this.onSave = this.onSave.bind(this);
  }

  componentDidMount() {
    // eslint-disable-next-line no-shadow
    const { allDetails, getUserDetails } = this.props;
    const payload = {
      id: allDetails.id,
    };
    if (allDetails && allDetails.id) {
      getUserDetails(payload).then(res => {
        if (res.type === 'GET_USER_PROFILE_DETAILS_SUCCESS') {
          // eslint-disable-next-line react/destructuring-assignment
          const details = this.props.allDetails;
          this.setState({
            firstName: details.firstName,
            lastName: details.lastName,
            company: details.company,
            department: details.department,
            position: details.position,
            email: details.email,
            // eslint-disable-next-line react/no-unused-state
            id: details.id,
          });
        }
        return null;
      });
    }
  }

  onUpdate(key, value) {
    this.setState({ [key]: value }, () => {});
  }

  /**
   * @method onSave : Method called on form submition to store user profile details in reducer and database
   */
  onSave() {
    const { firstName, lastName, company, department, position, email } = this.state;
    if (email === '' || !email) {
      alert('Please enter the email', 'danger');
      return false;
    }
    if (!Validation.validateEmail(email)) {
      alert('Please enter correct email', 'danger');
      return false;
    }
    // eslint-disable-next-line no-shadow
    const { userProfileDetails } = this.props;
    const payload = {
      firstName,
      lastName,
      company,
      department,
      position,
      email,
    };
    // Function to call set user details API
    userProfileDetails(payload).then(() => {
      alert('User Details saved successfully', 'danger');
    });
    return true;
  }

  render() {
    const { firstName, lastName, company, department, position, email } = this.state;
    return (
      <div>
        <header className="headerOfUserProfile">
          <img src={logo} className="App-logo" alt="logo" />
        </header>
        <div className="userProfileDiv">
          <Card className="GridcolSize2 newRegisterPage disabled-card">
            <CardBody>
              <p className="pTag">Please fill in this form to create an account.</p>
            </CardBody>
          </Card>
          <div style={{ margin: '30px' }}>
            <form>
              <div>
                <Label>
                  <b>First Name</b>
                </Label>
                <input
                  className="inputField"
                  type="text"
                  placeholder="Enter First Name"
                  name="firstName"
                  value={firstName}
                  onChange={e => this.onUpdate('firstName', e.currentTarget.value)}
                />
                <Label>
                  <b>Last Name</b>
                </Label>
                <input
                  className="inputField"
                  type="text"
                  placeholder="Enter Last Name"
                  name="lastName"
                  value={lastName}
                  onChange={e => this.onUpdate('lastName', e.currentTarget.value)}
                />
                <Label>
                  <b>Company</b>
                </Label>
                <input
                  className="inputField"
                  type="text"
                  placeholder="Enter Company"
                  name="company"
                  value={company}
                  onChange={e => this.onUpdate('company', e.currentTarget.value)}
                />
                <Label>
                  <b>Department</b>
                </Label>
                <input
                  className="inputField"
                  type="text"
                  placeholder="Enter department"
                  name="department"
                  value={department}
                  onChange={e => this.onUpdate('department', e.currentTarget.value)}
                />
                <Label>
                  <b>Position</b>
                </Label>
                <input
                  className="inputField"
                  type="text"
                  placeholder="Enter Position"
                  name="position"
                  value={position}
                  onChange={e => this.onUpdate('position', e.currentTarget.value)}
                />
                <Label>
                  <b>Email</b> <span style={{ color: 'red' }}>*</span>
                </Label>
                <input
                  className="inputField"
                  type="text"
                  placeholder="Enter Email"
                  name="email"
                  value={email}
                  onChange={e => this.onUpdate('email', e.currentTarget.value)}
                />

                <div className="saveButton">
                  <Button className="submitOrCancel" onClick={() => this.onSave()}>
                    Save
                  </Button>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

ProfileScreen.propTypes = {
  userProfileDetails: PropTypes.func.isRequired,
  getUserDetails: PropTypes.func.isRequired,
  allDetails: PropTypes.oneOfType([PropTypes.object]).isRequired,
};

const mapStateToProps = state => ({
  allDetails: state.SetUserProfileDetailsReducer,
});

const mapDispatchToProps = {
  userProfileDetails,
  getUserDetails,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProfileScreen);
