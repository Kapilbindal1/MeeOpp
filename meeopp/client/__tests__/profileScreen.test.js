import React from 'react';
import renderer from 'react-test-renderer';

import { ProfileScreen } from '../src/views/profileScreen/index';
import { getUserDetails, userProfileDetails } from '../src/redux/userProfile/action';

describe('Profile Screen', () => {
  test('should render correctly', () => {
    const data = {
      firstName: 'John',
      lastName: 'William',
      company: 'innow8',
      department: 'Computer science',
      position: 'Developer',
      email: 'contact@innow8apps.com',
    };
    const testRenderer = renderer.create(
      <ProfileScreen
        allDetails={data}
        getUserDetails={getUserDetails}
        userProfileDetails={userProfileDetails}
      />
    );
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
