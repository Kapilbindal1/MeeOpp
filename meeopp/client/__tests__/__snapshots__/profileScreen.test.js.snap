// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Screen should render correctly 1`] = `
<div>
  <header
    className="headerOfUserProfile"
  >
    <img
      alt="logo"
      className="App-logo"
      src="test-file-stub"
    />
  </header>
  <div
    className="userProfileDiv"
  >
    <div
      className="GridcolSize2 newRegisterPage disabled-card card"
    >
      <div
        className="card-body"
      >
        <p
          className="pTag"
        >
          Please fill in this form to create an account.
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "margin": "30px",
        }
      }
    >
      <form>
        <div>
          <label
            className=""
          >
            <b>
              First Name
            </b>
          </label>
          <input
            className="inputField"
            name="firstName"
            onChange={[Function]}
            placeholder="Enter First Name"
            type="text"
            value="John"
          />
          <label
            className=""
          >
            <b>
              Last Name
            </b>
          </label>
          <input
            className="inputField"
            name="lastName"
            onChange={[Function]}
            placeholder="Enter Last Name"
            type="text"
            value="William"
          />
          <label
            className=""
          >
            <b>
              Company
            </b>
          </label>
          <input
            className="inputField"
            name="company"
            onChange={[Function]}
            placeholder="Enter Company"
            type="text"
            value="innow8"
          />
          <label
            className=""
          >
            <b>
              Department
            </b>
          </label>
          <input
            className="inputField"
            name="department"
            onChange={[Function]}
            placeholder="Enter department"
            type="text"
            value="Computer science"
          />
          <label
            className=""
          >
            <b>
              Position
            </b>
          </label>
          <input
            className="inputField"
            name="position"
            onChange={[Function]}
            placeholder="Enter Position"
            type="text"
            value="Developer"
          />
          <label
            className=""
          >
            <b>
              Email
            </b>
             
            <span
              style={
                Object {
                  "color": "red",
                }
              }
            >
              *
            </span>
          </label>
          <input
            className="inputField"
            name="email"
            onChange={[Function]}
            placeholder="Enter Email"
            type="text"
            value="contact@innow8apps.com"
          />
          <div
            className="saveButton"
          >
            <button
              aria-label={null}
              className="submitOrCancel btn btn-secondary"
              onClick={[Function]}
              type="button"
            >
              Save
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
